File: backend/app/schemas/product.py
"""
[IDENTITY]: Product Schemas
DTOs for Product Management.

[INPUT]:
- ProductCreate (requires `original_asset_id`), ProductUpdate.

[LINK]:
- ProductRouter -> ../api/v1/endpoints/products.py
- ProductModel -> ../models/product.py

[OUTPUT]: ProductResponse.
[POS]: /backend/app/schemas/product.py

[PROTOCOL]:
1. `ProductCreate` enforces minimal valid state (Name + Category + Asset).
"""

from pydantic import BaseModel, Field
from datetime import datetime
from uuid import UUID

from app.models.product import ProductCategory, ProductStatus


class ProductBase(BaseModel):
    """Base schema for Product data."""
    name: str = Field(..., min_length=1, max_length=255)
    category: ProductCategory = Field(..., description="Product category enum value")


class ProductCreate(ProductBase):
    """Schema for creating a new Product."""
    original_asset_id: UUID


class ProductUpdate(BaseModel):
    """Schema for updating a Product."""
    name: str | None = Field(None, min_length=1, max_length=255)
    category: ProductCategory | None = None


class ProductResponse(ProductBase):
    """Schema for Product API responses."""
    id: UUID
    workspace_id: UUID
    original_asset_id: UUID
    status: ProductStatus
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/product.ts
/**
 * === HIGH-INTEGRITY HEADER ===
 *
 * [IDENTITY]: Product Types
 * Frontend type definitions for product management.
 *
 * Story关联: Story 1.2 - Product Catalog
 *
 * [INPUT]: N/A (Type Definition)
 *
 * [LINK]:
 * - Backend Schemas -> backend/app/schemas/product.py
 * - API Client -> @/lib/api/products.ts
 *
 * [OUTPUT]:
 * - Product, ProductCategory, ProductStatus
 * - PRODUCT_CATEGORIES, isProductCategory()
 *
 * [POS]: /frontend/src/types/product.ts
 *
 * [PROTOCOL]:
 * 1. ProductStatus must align with Backend Enum.
 *
 * === END HEADER ===
 */

export type ProductCategory =
    | 'clothing'
    | 'electronics'
    | 'beauty'
    | 'home'
    | 'food'
    | 'sports'
    | 'toys'
    | 'books'
    | 'automotive'
    | 'health'
    | 'other';

export const PRODUCT_CATEGORIES: readonly ProductCategory[] = [
    'clothing',
    'electronics',
    'beauty',
    'home',
    'food',
    'sports',
    'toys',
    'books',
    'automotive',
    'health',
    'other',
];

const PRODUCT_CATEGORY_SET: ReadonlySet<string> = new Set(PRODUCT_CATEGORIES);

export function isProductCategory(value: unknown): value is ProductCategory {
    return typeof value === 'string' && PRODUCT_CATEGORY_SET.has(value);
}

export type ProductStatus =
    | 'draft'
    | 'ready'
    | 'processing'
    | 'completed'
    | 'archived';

export interface CategoryInfo {
    id: ProductCategory;
    label: string;
    icon: string; // Lucide icon name
    description?: string;
    gradient?: string; // For card styling
}

export interface Product {
    id: string;
    workspaceId: string;
    name: string;
    category: ProductCategory;
    description?: string;
    originalAssetId: string;
    status: ProductStatus;
    createdAt: string;
    updatedAt: string;
}

export interface ProductCreateRequest {
    name: string;
    category: ProductCategory;
    original_asset_id: string;
}

export interface ProductUpdateRequest {
    name?: string;
    category?: ProductCategory;
}
